libraries/AC_AutoTune/AC_AutoTune_Multi.h

```diff
@@ -0,0 +1,134 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  support for autotune of multirotors. Based on original autotune code from ArduCopter, written by Leonard Hall
+  Converted to a library by Andrew Tridgell
+ */
+
+#pragma once
+
+#include "AC_AutoTune.h"
+
+class AC_AutoTune_Multi : public AC_AutoTune
+{
+public:
+    // constructor
+    AC_AutoTune_Multi();
+
+    // save gained, called on disarm
+    void save_tuning_gains() override;
+
+    // var_info for holding Parameter information
+    static const struct AP_Param::GroupInfo var_info[];
+
+protected:
+    // load gains
+    void load_test_gains() override;
+
+    // get intra test rate I gain for the specified axis
+    float get_intra_test_ri(AxisType test_axis) override;
+
+    // get tuned rate I gain for the specified axis
+    float get_tuned_ri(AxisType test_axis) override;
+
+    // get tuned yaw rate D gain
+    float get_tuned_yaw_rd() override { return 0.0f; }
+
+    void test_init() override;
+    void test_run(AxisType test_axis, const float dir_sign) override;
+    void do_gcs_announcements() override;
+
+    // update gains for the rate P up tune type
+    void updating_rate_p_up_all(AxisType test_axis) override;
+
+    // update gains for the rate P down tune type
+    void updating_rate_p_down_all(AxisType test_axis) override {};
+
+    // update gains for the rate D up tune type
+    void updating_rate_d_up_all(AxisType test_axis) override;
+
+    // update gains for the rate D down tune type
+    void updating_rate_d_down_all(AxisType test_axis) override;
+
+    // update gains for the rate ff up tune type
+    void updating_rate_ff_up_all(AxisType test_axis) override {};
```

Comment:
I think it might be a good idea to add internal errors to the implementation of multicopter methods that we think should never be called.  This will add a little flash but it will also help us catch programming errors which are very possible because this is a fairly complex change.

Reply:
Not sure what you mean by this statement "add internal errors"

Reply:
```
        if (!wp_nav->get_wp_destination_loc(default_loc)) {
            // this should never happen
            INTERNAL_ERROR(AP_InternalError::error_t::flow_of_control);
        }
```

Reply:
done